### ISSUE TITLE
Is it possible to turn wtconv2d into 1D for depth separable convolution in time series？

### ISSUE THREAD

**Issue by firenzemyz** [BGU-CS-VIL/WTConv#6]
Thanks for your contribution! Is it possible to turn wtconv2d into 1D for depth separable convolution in time series？As well as 3d，thanks for your answer!

**Comment by shahaffind** [comment]
Thank you for your interest in our work :)
It should be possible, you just have to modify several things:
1. wtconv/util/wavelet.py
`create_wavelet_filter` should return a 1D/3D filter for the convolution of the WT (note that for 1D, it's actually simpler, as it is what you get from `pywt`, just wrap it with `torch.tensor`). Keep in mind that for 1D it should be 2 filters (just low and high), and for 3D it should be 8 filters (LLL, LLH, LHL, LHH, HLL, HLH, HHL, HHH)
2. on the same file
`wavelet_transform` should also be modified to work with the 1D/3D WT filters (keep in mind the number of frequency maps)
3. wtconv/wtconv2d.py
`self.base_conv` and `self.wavelet_convs` should use the 1D or 3D torch convs. Note that `self.wavelet_convs` and `self.wavelet_scale` (and possibly other spots in the forward function) should also have the correct channel multiplier (2 or 8, the number of WT filters).

I hope this will help you with your project :)
Please let me know if it works (so I can close this issue) or if it fails (so I can provide more help).

**Comment by firenzemyz** [comment]
Thank you for your help! Is there anything wrong with the following code?`import pywt
import torch
from torch import nn
from functools import partial
import torch.nn.functional as F

def create_wavelet_filter_1d(wave, in_size, type=torch.float):
    w = pywt.Wavelet(wave)
    dec_hi = torch.tensor(w.dec_hi[::-1], dtype=type)
    dec_lo = torch.tensor(w.dec_lo[::-1], dtype=type)
    dec_filters = torch.stack([dec_lo, dec_hi], dim=0).unsqueeze(1)
    dec_filters = dec_filters.repeat(in_size, 1, 1)
    return dec_filters

def wavelet_transform_1d(x, filters):
    b, c, l = x.shape
    pad = (filters.shape[2] // 2 - 1,)
    x = F.conv1d(x, filters, stride=2, groups=c, padding=pad)
    x = x.reshape(b, c, 2, l // 2)
    return x

class WTConv1d(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=5, stride=1, bias=True, wt_levels=1, wt_type='db1'):
        super(WTConv1d, self).__init__()

        assert in_channels == out_channels

        self.in_channels = in_channels
        self.wt_levels = wt_levels
        self.stride = stride

        self.wt_filter = create_wavelet_filter_1d(wt_type, in_channels, torch.float)
        self.wt_filter = nn.Parameter(self.wt_filter, requires_grad=False)

        self.wt_function = partial(wavelet_transform_1d, filters=self.wt_filter)

        self.base_conv = nn.Conv1d(in_channels, in_channels, kernel_size, padding='same', stride=1, bias=bias)
        self.base_scale = _ScaleModule([1, in_channels, 1])

        self.wavelet_convs = nn.ModuleList(
            [nn.Conv1d(in_channels * 2, in_channels * 2, kernel_size, padding='same', stride=1, bias=False)
             for _ in range(self.wt_levels)]
        )
        self.wavelet_scale = nn.ModuleList(
            [_ScaleModule([1, in_channels * 2, 1], init_scale=0.1) for _ in range(self.wt_levels)]
        )

        if self.stride > 1:
            self.stride_filter = nn.Parameter(torch.ones(in_channels, 1, 1), requires_grad=False)
            self.do_stride = lambda x_in: F.conv1d(x_in, self.stride_filter, bias=None, stride=self.stride, groups=in_channels)
        else:
            self.do_stride = None

    def forward(self, x):
        x_ll_in_levels = []
        x_h_in_levels = []
        shapes_in_levels = []

        curr_x_ll = x

        for i in range(self.wt_levels):
            curr_shape = curr_x_ll.shape
            shapes_in_levels.append(curr_shape)
            if curr_shape[2] % 2 > 0:
                curr_pads = (0, curr_shape[2] % 2)
                curr_x_ll = F.pad(curr_x_ll, curr_pads)

            curr_x = self.wt_function(curr_x_ll)
            curr_x_ll = curr_x[:, :, 0, :]

            shape_x = curr_x.shape
            curr_x_tag = curr_x.reshape(shape_x[0], shape_x[1] * 2, shape_x[3])
            curr_x_tag = self.wavelet_scale[i](self.wavelet_convs[i](curr_x_tag))
            curr_x_tag = curr_x_tag.reshape(shape_x)

            x_ll_in_levels.append(curr_x_tag[:, :, 0, :])
            x_h_in_levels.append(curr_x_tag[:, :, 1, :])

        next_x_ll = 0

        for i in range(self.wt_levels - 1, -1, -1):
            curr_x_ll = x_ll_in_levels.pop()
            curr_x_h = x_h_in_levels.pop()
            curr_shape = shapes_in_levels.pop()

            curr_x_ll = curr_x_ll + next_x_ll

            curr_x = torch.cat([curr_x_ll.unsqueeze(2), curr_x_h.unsqueeze(2)], dim=2)
            next_x_ll = F.conv_transpose1d(curr_x.view(curr_x.shape[0], -1, curr_x.shape[3]), self.wt_filter, stride=2, groups=self.in_channels)
            next_x_ll = next_x_ll[:, :, :curr_shape[2]]

        x_tag = next_x_ll
        assert len(x_ll_in_levels) == 0

        x = self.base_scale(self.base_conv(x))
        x = x + x_tag

        if self.do_stride is not None:
            x = self.do_stride(x)

        return x

class _ScaleModule(nn.Module):
    def __init__(self, dims, init_scale=1.0, init_bias=0):
        super(_ScaleModule, self).__init__()
        self.dims = dims
        self.weight = nn.Parameter(torch.ones(*dims) * init_scale)
        self.bias = None

    def forward(self, x):
        return torch.mul(self.weight, x)

class DepthwiseSeparableConvWithWTConv1d(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=3):
        super(DepthwiseSeparableConvWithWTConv1d, self).__init__()

        self.depthwise = WTConv1d(in_channels, in_channels, kernel_size=kernel_size)
        self.pointwise = nn.Conv1d(in_channels, out_channels, kernel_size=1, stride=1, padding=0, bias=False)

    def forward(self, x):
        x = self.depthwise(x)
        x = self.pointwise(x)
        return x

if __name__ == '__main__':
    input_data = torch.randn(1, 64, 256)  # Example input data (batch_size, channels, length)
    wtconv = DepthwiseSeparableConvWithWTConv1d(in_channels=64, out_channels=128)
    output_data = wtconv(input_data)
    print("Input shape:", input_data.shape)
    print("Output shape:", output_data.shape)
`

**Comment by shahaffind** [comment]
Note that you need the rec and dec filters from pywt if you're using other wavelet basis then Haar (they can be different from each other, i.e. with biorthogonal basis)

Otherwise seems fine. Let me know if it works :)

**Comment by firenzemyz** [comment]
According to your reply, I updated the code again, thank you very much!
```python
#1D
import pywt
import torch
from torch import nn
from functools import partial
import torch.nn.functional as F

def create_wavelet_filter_1d(wave, in_size, type=torch.float):
    w = pywt.Wavelet(wave)
    dec_hi = torch.tensor(w.dec_hi[::-1], dtype=type)
    dec_lo = torch.tensor(w.dec_lo[::-1], dtype=type)
    rec_hi = torch.tensor(w.rec_hi[::-1], dtype=type)
    rec_lo = torch.tensor(w.rec_lo[::-1], dtype=type)

    # Decomposition filters
    dec_filters = torch.stack([dec_lo, dec_hi], dim=0)
    dec_filters = dec_filters[:, None].repeat(in_size, 1, 1)

    # Reconstruction filters
    rec_filters = torch.stack([rec_lo, rec_hi], dim=0)
    rec_filters = rec_filters[:, None].repeat(in_size, 1, 1)

    return dec_filters, rec_filters

def wavelet_transform_1d(x, filters):
    b, c, l = x.shape
    pad = filters.shape[2] // 2 - 1
    x = F.conv1d(x, filters, stride=2, groups=c, padding=pad)
    x = x.reshape(b, c, 2, l // 2)
    return x

def inverse_wavelet_transform_1d(x, filters):
    b, c, _, l_half = x.shape
    pad = filters.shape[2] // 2 - 1
    x = x.reshape(b, c * 2, l_half)
    x = F.conv_transpose1d(x, filters, stride=2, groups=c, padding=pad)
    return x

class WTConv1d(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=5, stride=1, bias=True, wt_levels=1, wt_type='db1'):
        super(WTConv1d, self).__init__()

        assert in_channels == out_channels

        self.in_channels = in_channels
        self.wt_levels = wt_levels
        self.stride = stride

        self.wt_filter, self.iwt_filter = create_wavelet_filter_1d(wt_type, in_channels, torch.float)
        self.wt_filter = nn.Parameter(self.wt_filter, requires_grad=False)
        self.iwt_filter = nn.Parameter(self.iwt_filter, requires_grad=False)

        self.wt_function = partial(wavelet_transform_1d, filters=self.wt_filter)
        self.iwt_function = partial(inverse_wavelet_transform_1d, filters=self.iwt_filter)

        self.base_conv = nn.Conv1d(in_channels, in_channels, kernel_size, padding='same', stride=1, bias=bias)
        self.base_scale = _ScaleModule([1, in_channels, 1])

        self.wavelet_convs = nn.ModuleList(
            [nn.Conv1d(in_channels * 2, in_channels * 2, kernel_size, padding='same', stride=1, bias=False)
             for _ in range(self.wt_levels)]
        )
        self.wavelet_scale = nn.ModuleList(
            [_ScaleModule([1, in_channels * 2, 1], init_scale=0.1) for _ in range(self.wt_levels)]
        )

        if self.stride > 1:
            self.stride_filter = nn.Parameter(torch.ones(in_channels, 1, 1), requires_grad=False)
            self.do_stride = lambda x_in: F.conv1d(x_in, self.stride_filter, bias=None, stride=self.stride, groups=in_channels)
        else:
            self.do_stride = None

    def forward(self, x):
        x_ll_in_levels = []
        x_h_in_levels = []
        shapes_in_levels = []

        curr_x_ll = x

        for i in range(self.wt_levels):
            curr_shape = curr_x_ll.shape
            shapes_in_levels.append(curr_shape)
            if curr_shape[2] % 2 > 0:
                curr_pads = (0, curr_shape[2] % 2)
                curr_x_ll = F.pad(curr_x_ll, curr_pads)

            curr_x = self.wt_function(curr_x_ll)
            curr_x_ll = curr_x[:, :, 0, :]

            shape_x = curr_x.shape
            curr_x_tag = curr_x.reshape(shape_x[0], shape_x[1] * 2, shape_x[3])
            curr_x_tag = self.wavelet_scale[i](self.wavelet_convs[i](curr_x_tag))
            curr_x_tag = curr_x_tag.reshape(shape_x)

            x_ll_in_levels.append(curr_x_tag[:, :, 0, :])
            x_h_in_levels.append(curr_x_tag[:, :, 1, :])

        next_x_ll = 0

        for i in range(self.wt_levels - 1, -1, -1):
            curr_x_ll = x_ll_in_levels.pop()
            curr_x_h = x_h_in_levels.pop()
            curr_shape = shapes_in_levels.pop()

            curr_x_ll = curr_x_ll + next_x_ll

            curr_x = torch.cat([curr_x_ll.unsqueeze(2), curr_x_h.unsqueeze(2)], dim=2)
            next_x_ll = self.iwt_function(curr_x)

            next_x_ll = next_x_ll[:, :, :curr_shape[2]]

        x_tag = next_x_ll
        assert len(x_ll_in_levels) == 0

        x = self.base_scale(self.base_conv(x))
        x = x + x_tag

        if self.do_stride is not None:
            x = self.do_stride(x)

        return x

class _ScaleModule(nn.Module):
    def __init__(self, dims, init_scale=1.0, init_bias=0):
        super(_ScaleModule, self).__init__()
        self.dims = dims
        self.weight = nn.Parameter(torch.ones(*dims) * init_scale)
        self.bias = None

    def forward(self, x):
        return torch.mul(self.weight, x)

class DepthwiseSeparableConvWithWTConv1d(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=3):
        super(DepthwiseSeparableConvWithWTConv1d, self).__init__()

        self.depthwise = WTConv1d(in_channels, in_channels, kernel_size=kernel_size)
        self.pointwise = nn.Conv1d(in_channels, out_channels, kernel_size=1, stride=1, padding=0, bias=False)

    def forward(self, x):
        x = self.depthwise(x)
        x = self.pointwise(x)
        return x

if __name__ == '__main__':
    input_data = torch.randn(1, 64, 1024)  # Example input data (batch_size, channels, length)
    wtconv = DepthwiseSeparableConvWithWTConv1d(in_channels=64, out_channels=128)
    output_data = wtconv(input_data)
    print("Input shape:", input_data.shape)
    print("Output shape:", output_data.shape)
```

**Comment by shahaffind** [comment]
No problem :)
I am closing the issue for now. Keep me posted if you have some interesting results or more questions :)
